WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift

```diff
@@ -51,23 +56,315 @@ import Foundation
+        postUploadOp.currentStatus = .InProgress
+        coreDataStack.saveContext()
+
+        remote.createPost(postUploadOp.remotePost, success: { post in
+            if let post = post {
+                if let postID = post.postID {
+                    DDLogInfo("Post \(post.postID.stringValue) sucessfully uploaded to site \(post.siteID.stringValue)")
+                    postUploadOp.remotePostID = postID.int64Value
+                }
+                postUploadOp.currentStatus = .Complete
+                self.coreDataStack.saveContext()
+                self.cleanupSessionAndTerminate()
+            }
+        }, failure: { error in
+            var errorString = "Error creating post"
+            if let error = error as NSError? {
+                errorString += ": \(error.localizedDescription)"
+            }
+            self.logError(errorString: errorString, uploadOpObjectIDs: postUploadOpID)
+            self.cleanupSessionAndTerminate()
+        })
+    }
 }
 
-extension ShareExtensionSessionManager: URLSessionTaskDelegate {
+// MARK: - ShareExtensionSessionManager Extension: Persistence Helpers
+
+private extension ShareExtensionSessionManager {
+
+    /// Updates the status using the given uploadOp's ObjectID.
+    ///
+    /// - Parameters:
+    ///   - status: New status
+    ///   - uploadOpObjectID: Managed object ID for a given upload op
+    ///
+    func updateStatus(_ status: UploadOperation.UploadStatus, forUploadOpWithObjectID uploadOpObjectID: NSManagedObjectID) {
+        var uploadOp: UploadOperation?
+        do {
+            uploadOp = try coreDataStack.managedContext.existingObject(with: uploadOpObjectID) as? UploadOperation
+        } catch {
+            DDLogError("Error setting \(status.stringValue()) status for UploadOperation Object with ID: \(uploadOpObjectID) — could not fetch object.")
+            return
+        }
+        uploadOp?.currentStatus = status
+        coreDataStack.saveContext()
+    }
+
+    /// Updates the remote media URL and remote media ID on an upload op that corresponds with the provided
+    /// file name. If a parameter is nil, that specific param will not be updated.
+    ///
+    /// Note: We are searching for the upload op using a filename because a given task ID can have
+    /// multiple files associated with it.
+    ///
+    /// - Parameters:
+    ///   - fileName: the fileName from a URLSessionTask
+    ///   - remoteMediaID: remote media ID. Can be nil.
+    ///   - remoteURL: remote media URL string. Can be nil.
+    func updateMediaOperationForFileName(_ fileName: String, remoteMediaID: Int64?, remoteURL: String?) {
+        guard let uploadMediaOp = fetchUploadOpForFileName(fileName) else {
+            DDLogError("Error loading UploadOperation Object with File Name: \(fileName)")
+            return
+        }
+
+        if let remoteID = remoteMediaID {
```

Comment:
Do we need to unwrap these before we assign them? The property in the core data is optional, so could we update the Swift property declaration to be optional too?

One other note – When doing unwrapping like this, you can use the same variable name: 

`if let remoteMediaID = remoteMediaID {` for example :)